middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -574,47 +495,81 @@ let split_complex_binding ~env ~res (binding : complex binding) =
         inline = binding.inline;
         bound_expr = Split { cmm_expr = new_cmm_expr };
         cmm_var = binding.cmm_var
       }
     in
     res, Split { new_bindings; split_binding }
 
-let remove_binding env var =
-  { env with bindings = Variable.Map.remove var env.bindings }
-
-let will_inline_simple env res { effs; bound_expr = Simple { cmm_expr }; _ } =
-  cmm_expr, env, res, effs
-
-let will_inline_complex env res { effs; bound_expr; _ } =
-  match bound_expr with
-  | Split { cmm_expr } -> cmm_expr, env, res, effs
-  | Splittable_prim { dbg; prim; args } ->
-    let cmm_expr, res = rebuild_prim ~dbg ~env ~res prim (List.map fst args) in
-    cmm_expr, env, res, effs
+(* Adding binding to the env and split them *)
 
-let will_not_inline_simple env res { cmm_var; bound_expr = Simple _; _ } =
-  ( C.var (Backend_var.With_provenance.var cmm_var),
-    env,
-    res,
-    Ece.pure_can_be_duplicated )
+let rec add_binding_to_env ?extra env res var (Binding binding as b) =
+  let env =
+    let bindings = Variable.Map.add var b env.bindings in
+    let cmm_expr = C.var (Backend_var.With_provenance.var binding.cmm_var) in
+    let vars = Variable.Map.add var cmm_expr env.vars in
+    let vars_extra =
+      match extra with
+      | None -> env.vars_extra
+      | Some info -> Variable.Map.add var info env.vars_extra
+    in
+    { env with bindings; vars; vars_extra }
+  in
+  let classification =
+    To_cmm_effects.classify_by_effects_and_coeffects binding.effs
+  in
+  let inline : _ inline = binding.inline in
+  match inline with
+  | Must_inline_and_duplicate -> (
+    match classification with
+    | Pure | Generative_immutable -> env, res
+    | Coeffect_only | Effect ->
+      let env, res, _ = split_in_env env res var (binding : complex binding) in
+      env, res)
+  | May_inline_once | Must_inline_once | Do_not_inline -> (
+    match classification with
+    | Pure -> env, res
+    | Generative_immutable -> (
+      match inline with
+      | Must_inline_once -> env, res
+      | May_inline_once | Do_not_inline ->
+        (* Generative expressions not marked as `must_inline` are treated as
+           having effects, since function from the `Gc` module can read counters
+           that are increased by allocations. *)
+        { env with stages = Effect var :: env.stages }, res
+      | Must_inline_and_duplicate -> assert false (* impossible to reach *))
+    | Effect -> { env with stages = Effect var :: env.stages }, res
+    | Coeffect_only ->
+      let stages =
+        match env.stages with
+        | Coeffect_only vars :: stages ->
+          (* Multiple coeffect-only bindings may be accumulated in the same
+             stage. *)
+          Coeffect_only (Variable.Set.add var vars) :: stages
+        | [] | Effect _ :: _ ->
+          Coeffect_only (Variable.Set.singleton var) :: env.stages
+      in
+      { env with stages }, res)
 
-let split_and_inline env res var binding =
+and split_in_env env res var binding =
```

Comment:
The name is not very good. Please leave a CR so that we can try to find a better name later.

Reply:
done